service/src/test/java/bio/terra/workspace/service/workspace/ApplicationServiceTest.java

```diff
@@ -0,0 +1,212 @@
+
+  @Autowired ApplicationDao appDao;
+  @Autowired WsmApplicationService appService;
+  @Autowired WorkspaceService workspaceService;
+  @Autowired JobService jobService;
+
+  /** Mock SamService does nothing for all calls that would throw if unauthorized. */
+  @MockBean private SamService mockSamService;
+
+  private UUID workspaceId;
+  private UUID workspaceId2;
+
+  @BeforeEach
+  void setup() throws Exception {
+    // Set up so all spend profile and workspace checks are successful
+    Mockito.when(
+            mockSamService.isAuthorized(
+                Mockito.any(),
+                Mockito.eq(SamConstants.SPEND_PROFILE_RESOURCE),
+                Mockito.any(),
+                Mockito.eq(SamConstants.SPEND_PROFILE_LINK_ACTION)))
+        .thenReturn(true);
+    Mockito.when(
+            mockSamService.isAuthorized(
+                Mockito.any(),
+                Mockito.eq(SamConstants.SAM_WORKSPACE_RESOURCE),
+                Mockito.any(),
+                Mockito.any()))
+        .thenReturn(true);
+
+    // Populate the applications - this should be idempotent since we are
+    // re-creating the same configuration every time.
+    Map<UUID, WsmDbApplication> dbAppMap = appService.buildAppMap();
+    appService.processApp(LEO_APP, dbAppMap);
+    appService.processApp(CARMEN_APP, dbAppMap);
+    appService.processApp(NORM_APP, dbAppMap);
+
+    // Create two workspaces
+    workspaceId = UUID.randomUUID();
+    var request =
+        WorkspaceRequest.builder()
+            .workspaceId(workspaceId)
+            .spendProfileId(Optional.empty())
+            .workspaceStage(WorkspaceStage.MC_WORKSPACE)
+            .build();
+
+    workspaceService.createWorkspace(request, USER_REQUEST);
+
+    workspaceId2 = UUID.randomUUID();
+    request =
+        WorkspaceRequest.builder()
+            .workspaceId(workspaceId2)
+            .spendProfileId(Optional.empty())
+            .workspaceStage(WorkspaceStage.MC_WORKSPACE)
+            .build();
+    workspaceService.createWorkspace(request, USER_REQUEST);
+  }
+
+  @Test
+  public void applicationEnableTest() {
+    // Verify no apps enabled in workspace
+    enumerateCheck(false, false, false);
+
+    // enable leo - should work
+    appService.enableWorkspaceApplication(USER_REQUEST, workspaceId, LEO_UUID);
+
+    // enable carmen - should fail - deprecated
+    assertThrows(
+        InvalidApplicationStateException.class,
+        () -> appService.enableWorkspaceApplication(USER_REQUEST, workspaceId, CARMEN_UUID));
+
+    // enable norm - should fail - decommissioned
+    assertThrows(
+        InvalidApplicationStateException.class,
+        () -> appService.enableWorkspaceApplication(USER_REQUEST, workspaceId, NORM_UUID));
+
+    enumerateCheck(true, false, false);
+
+    // Use the DAO directly and skip the check to enable carmen so we can check that state.
+    appDao.enableWorkspaceApplicationNoCheck(workspaceId, CARMEN_UUID);
+    enumerateCheck(true, true, false);
```

Comment:
Not sure I'm following here -- when do you want to skip the check in normal operation?